softwareComponents/kinematics/include/aabb.hpp

```diff
@@ -0,0 +1,227 @@
+#pragma once
+
+#include <optional>
+
+#include "shapes.hpp"
+
+
+namespace rofi::geometry {
+
+using namespace rofi::configuration::matrices;
+
+template< typename T >
+concept geometric = requires ( const Box& box, const T& shape ){
+    collide( box, shape );
+    shape.bounding_box();
+};
+
+// Leaves are expected to be spheres or polygons
+template< geometric LeafShape >
+struct AABB_Node {
+
+    Box bound;
+
+    // Node is either a leaf with a shape or has some children
+    std::optional< LeafShape > shape;
+    std::array< std::unique_ptr< AABB_Node >, 2 > children;
```

Comment:
Nebylo by idiomatičtější mít `AABB_InnerNode` a `AABB_Leaf` a `children` si držet jako kontejner a `std::variant< AABB_InnerNode, AABB_Leaf`? Proč si to myslím: Vnitřní uzly mají jiné metody než listy. A tedy volat `leftChild` a `rightChild` na uzlu, který je listem je nedefinované chování. Když budu mít variant, budu ho muset visitovat a tedy můžu mít čistší rozhraní pro dva typy uzů.